### PR TITLE
Wasm threads blog post

### LINKED ISSUE #5355: WebAssembly threads [6/14/2021]
* [Draft](https://docs.google.com/document/d/1yRtj4TftKzslSqy8N86XQsyNABjEHVDeGRL7mjQzJ-I/edit?usp=sharing)

### PR DESCRIPTION
**Pull request by RReverser**
Fixes #5355.

### CODE REVIEW

File: src/site/content/en/blog/webassembly-threads/index.md [hunk 1]
```diff
@@ -0,0 +1,472 @@
...
+initialization is similar to the `-s PTHREAD_POOL_SIZE=...` option in Emscripten explained earlier,
+and needs to happen before the main code to avoid the same deadlocks.
+
+There are few other configurations to set up before you can use multithreading, so make sure to
+check the [detailed documentation](https://github.com/GoogleChromeLabs/wasm-bindgen-rayon#usage) for
+complete setup steps.
+
+{% Aside 'caution' %} The same
+[caveats](https://github.com/GoogleChromeLabs/wasm-bindgen-rayon#caveats) about blocking the main
+thread apply here too. {% endAside %}
```
Comment (rachelandrew) [inline src/site/content/en/blog/webassembly-threads/index.md:428]: This is another caution that leads into the text, the same comment as above applies.

File: src/site/content/en/blog/webassembly-threads/index.md [hunk 2]
```diff
@@ -0,0 +1,472 @@
+---
+title: Using WebAssembly threads from C, C++ and Rust
```
Comment (mihajlija) [inline src/site/content/en/blog/webassembly-threads/index.md:2]:
```suggestion
title: Using WebAssembly threads from C, C++ and Rust
subhead: Learn how to bring multithreaded applications written in other languages to WebAssembly.
```

File: src/site/content/en/blog/webassembly-threads/index.md [hunk 3]
```diff
@@ -0,0 +1,472 @@
...
+  - performance
+---
+
+WebAssembly threads support is one of the most important performance additions to WebAssembly. It
+allows you to either run parts of your code in parallel on separate cores, or the same code over
+independent parts of the input data, scaling it to as many cores as the user has and significantly
+reducing the overall execution time.
+
+In this article you will learn how to use WebAssembly threads to bring multithreaded applications
+written in languages like C, C++ and Rust to the Web.
```
[the remark this answers is not in the document]
Comment (mihajlija) [reply src/site/content/en/blog/webassembly-threads/index.md:21]: @rachelandrew The "Web" shouldn't be capitalized, refer to the word list https://web.dev/handbook/word-list/ and avoid unnecessary capitalization in general.

File: src/site/content/en/blog/webassembly-threads/index.md [hunk 4]
```diff
@@ -0,0 +1,472 @@
...
+
+## How WebAssembly threads work
+
+WebAssembly threads is not a separate feature, but a combination of several components that allows
+WebAssembly apps to use traditional multithreading paradigms on the Web.
```
Comment (mihajlija) [inline src/site/content/en/blog/webassembly-threads/index.md:26]:
```suggestion
WebAssembly apps to use traditional multithreading paradigms on the web.
```

File: src/site/content/en/blog/webassembly-threads/index.md [hunk 5]
```diff
@@ -0,0 +1,472 @@
...
+[`WebAssembly.Module`](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/WebAssembly/Module)
+as well as a shared
+[`WebAssembly.Memory`](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/WebAssembly/Memory)
+(see below) with those other threads. This establishes communication and allows all those threads to
+run the same WebAssembly code on the same shared memory without going through JavaScript again.
+
+Web Workers have been around for over a decade now, are widely supported, and don't require any
+special flags.
+
+### SharedArrayBuffer
```
Comment (mihajlija) [inline src/site/content/en/blog/webassembly-threads/index.md:45]:
```suggestion
### `SharedArrayBuffer`
```

File: src/site/content/en/blog/webassembly-threads/index.md [hunk 6]
```diff
@@ -0,0 +1,472 @@
...
+    pthread_join(thread_id, NULL);
+
+    puts("After the thread");
+
+    return 0;
+}
+```
+
+Here the headers for the pthread library are included via `pthread.h`. You can also see a couple of
+crucial functions for dealing with threads.
```
Comment (mihajlija) [inline src/site/content/en/blog/webassembly-threads/index.md:200]:
```suggestion
Here the headers for the `pthread` library are included via `pthread.h`. You can also see a couple of
crucial functions for dealing with threads.
```

File: src/site/content/en/blog/webassembly-threads/index.md [hunk 7]
```diff
@@ -0,0 +1,472 @@
...
+
+Now let's take a look at how to build a multithreaded version of the WebAssembly module.
+
+### C
+
+In C, particularly on Unix-like systems, the common way to use threads is via [POSIX
+Threads](https://en.wikipedia.org/wiki/POSIX_Threads) provided by the `pthread` library. Emscripten
+[provides an API-compatible implementation](https://emscripten.org/docs/porting/pthreads.html) of
+the `pthread` library built atop Web Workers, shared memory and atomics, so that the same code can
+work on the Web without changes.
```
Comment (mihajlija) [inline src/site/content/en/blog/webassembly-threads/index.md:167]:
```suggestion
work on the web without changes.
```

File: src/site/content/en/blog/webassembly-threads/index.md [hunk 8]
```diff
@@ -0,0 +1,472 @@
...
+
+Here the headers for the pthread library are included via `pthread.h`. You can also see a couple of
+crucial functions for dealing with threads.
+
+[`pthread_create`](https://man7.org/linux/man-pages/man3/pthread_create.3.html) will create a
+background thread. It takes a destination to store a thread handle in, some thread creation
+attributes (here not passing any, so it's just `NULL`), the callback to be executed in the new
+thread (here `thread_callback`), and an optional argument pointer to pass to that callback in case
+you want to share some data from the main thread - in this example we're sharing a pointer to a
+variable `arg`.
```
[the remark this answers is not in the document]
Comment (mihajlija) [reply src/site/content/en/blog/webassembly-threads/index.md:207]: + 1, it's difficult to follow as is.

File: src/site/content/en/blog/webassembly-threads/index.md [hunk 9]
```diff
@@ -0,0 +1,472 @@
...
+
+[`pthread_join`](https://man7.org/linux/man-pages/man3/pthread_join.3.html) can be called later at
+any time to wait for the thread to finish the execution, and get the result returned from the
+callback. It accepts the earlier assigned thread handle as well as a pointer to store the result in
+- in this case I don't have any results, so I'm just passing `NULL`.
```
[the remark this answers is not in the document]
Comment (mihajlija) [reply src/site/content/en/blog/webassembly-threads/index.md:212]: This is not a rule we enforce strictly, but try to avoid speaking in first person in code instructions (both "we" and "I"). For example, you could rephrase like this: "In this case, there aren't any results so the function takes a `NULL` argument".

File: src/site/content/en/blog/webassembly-threads/index.md [hunk 10]
```diff
@@ -0,0 +1,472 @@
...
+Tried to spawn a new thread, but the thread pool is exhausted.
+This might result in a deadlock unless some threads eventually exit or the code
+explicitly breaks out to the event loop.
+If you want to increase the pool size, use setting `-s PTHREAD_POOL_SIZE=...`.
+If you want to throw an explicit error instead of the risk of deadlocking in those
+cases, use setting `-s PTHREAD_POOL_SIZE_STRICT=2`.
+[…hangs here…]
+```
+
+What happened? The problem is, most of the time-consuming APIs on the Web are asynchronous and rely
```
Comment (mihajlija) [inline src/site/content/en/blog/webassembly-threads/index.md:235]:
```suggestion
What happened? The problem is, most of the time-consuming APIs on the web are asynchronous and rely
```

File: src/site/content/en/blog/webassembly-threads/index.md [hunk 11]
```diff
@@ -0,0 +1,472 @@
...
+can scale to an arbitrary number of threads.
+
+In the example above, there is only one thread being created, so instead of reserving all cores it's
+sufficient to use `-s PTHREAD_POOL_SIZE=1`:
+
+```shell
+emcc -pthread -s PTHREAD_POOL_SIZE=1 example.c -o example.js
+```
+
+{% Aside 'success' %} This time, when you execute it, things work successfully: {% endAside %}
```
Comment (mihajlija) [inline src/site/content/en/blog/webassembly-threads/index.md:269]: Could you make this aside more descriptive? Something along the lines "When you execute [action] in a [correct way to do it] a [good thing happens (or a bad thing is avoided)]."

I'd also suggest to use "caution" or "gotcha" formatting around the previous example to contrast them.

File: src/site/content/en/blog/webassembly-threads/index.md [hunk 12]
```diff
@@ -0,0 +1,472 @@
...
+
+[`pthread_join`](https://man7.org/linux/man-pages/man3/pthread_join.3.html) can be called later at
+any time to wait for the thread to finish the execution, and get the result returned from the
+callback. It accepts the earlier assigned thread handle as well as a pointer to store the result in
+- in this case I don't have any results, so I'm just passing `NULL`.
+
+To compile code using threads with Emscripten, you need to invoke `emcc` and pass a `-pthread`
+parameter like you would when compiling same code with Clang or GCC on other platforms:
+
+```shell
```
Comment (mihajlija) [inline src/site/content/en/blog/webassembly-threads/index.md:217]: Use some of the formatting options to emphasise that this is not the best way to do it. https://web.dev/handbook/web-dev-components/#asides

File: src/site/content/en/blog/webassembly-threads/index.md [hunk 13]
```diff
@@ -0,0 +1,472 @@
...
+attributes (here not passing any, so it's just `NULL`), the callback to be executed in the new
+thread (here `thread_callback`), and an optional argument pointer to pass to that callback in case
+you want to share some data from the main thread - in this example we're sharing a pointer to a
+variable `arg`.
+
+[`pthread_join`](https://man7.org/linux/man-pages/man3/pthread_join.3.html) can be called later at
+any time to wait for the thread to finish the execution, and get the result returned from the
+callback. It accepts the earlier assigned thread handle as well as a pointer to store the result in
+- in this case I don't have any results, so I'm just passing `NULL`.
+
```
Comment (mihajlija) [inline src/site/content/en/blog/webassembly-threads/index.md:213]: Consider adding another heading level here, for example "Compiling C code with Emscripten".

File: src/site/content/en/blog/webassembly-threads/index.md [hunk 14]
```diff
@@ -0,0 +1,472 @@
...
+```text
+Before the thread
+Inside the thread: 42
+After the thread
+Pthread 0x701510 exited.
+```
+
+{% Aside 'caution' %} There is another problem though: see that `sleep(1)` in the code example? It
+executes in the thread callback, meaning off the main thread, so it should be fine, right? Well, it
+isn't. {% endAside %}
```
[the remark this answers is not in the document]
Comment (mihajlija) [reply src/site/content/en/blog/webassembly-threads/index.md:281]: Agreed. Another thing that's not great is that the aside is referring to a line in a code example that's a couple of scrolls above. It's a lot to expect from readers to keep in working memory while reading and it's potentially disorienting to scroll back to refer to that place in code. 

Suggestion: What Rachel said + repeat that function `void *thread_callback` and highlight the relevant line. Consider adding another subheading level for that.

File: src/site/content/en/blog/webassembly-threads/index.md [hunk 15]
```diff
@@ -0,0 +1,472 @@
...
+
+When `pthread_join` is called, it has to wait for the thread execution to finish, meaning that if
+the created thread is performing long-running tasks—in this case, sleeping 1 second—then the main
+thread will also have to block for the same amount of time till the results are back. When this JS
+is executed in the browser, it will block the UI thread for 1 second until the thread callback
+returns. This leads to poor user experience.
+
+There are few solutions to this.
```
Comment (mihajlija) [inline src/site/content/en/blog/webassembly-threads/index.md:289]: Can you briefly introduce the solutions in a list? It can be as simple as:
- `pthread_detach`
- `-s PROXY_TO_PTHREAD`
- create your own Worker and expose it with Comlink

File: src/site/content/en/blog/webassembly-threads/index.md [hunk 16]
```diff
@@ -0,0 +1,472 @@
...
+  numbers
+  .iter()
+  .par_iter()
+  .map(x => x * x)
+  .sum()
+}
+```
+
+This unassuming code will split up the input data, calculate `x * x` and partial sums in parallel
+threads, and in the end add up those partial results together.
```
Comment (mihajlija) [inline src/site/content/en/blog/webassembly-threads/index.md:395]:
```suggestion
This code will split up the input data, calculate `x * x` and partial sums in parallel
threads, and in the end add up those partial results together.
```

Comment (mihajlija) [reply src/site/content/en/blog/webassembly-threads/index.md:395]: Unnecessary adjective.

File: src/site/content/en/blog/webassembly-threads/index.md [hunk 17]
```diff
@@ -0,0 +1,472 @@
...
+
+In order to accomodate for platforms without working `std::thread`, Rayon provides hooks that allow
```
Comment (mihajlija) [inline src/site/content/en/blog/webassembly-threads/index.md:397]:
```suggestion
To accommodate for platforms without working `std::thread`, Rayon provides hooks that allow
```

File: src/site/content/en/blog/webassembly-threads/index.md [hunk 18]
```diff
@@ -0,0 +1,472 @@
...
+to define custom logic for spawning and exiting threads. We've tapped into those hooks and built an
```
[the remark this answers is not in the document]
Comment (mihajlija) [reply src/site/content/en/blog/webassembly-threads/index.md:398]: It's okay to use "we" here, but you need to define who "we" are.

File: src/site/content/en/blog/webassembly-threads/index.md [hunk 19]
```diff
@@ -0,0 +1,472 @@
...
+
+## Real-world use-cases
+
+We actively use WebAssembly threads in [Squoosh.app](https://squoosh.app/) for client-side image
+compression - in particular, for formats like AVIF (C++), JPEG-XL (C++), OxiPNG (Rust) and WebP v2
+(C++). Thanks to multithreading alone, we've seen consistent 1.5x-3x speed-ups (exact ratio differs
+per codec), and were able to push those numbers even further by combining WebAssembly threads with
+[WebAssembly SIMD](https://v8.dev/features/simd)!
+
+Google Earth is another notable service that's using WebAssembly threads for its [Web
```
Comment (mihajlija) [inline src/site/content/en/blog/webassembly-threads/index.md:464]:
```suggestion
Google Earth is another notable service that's using WebAssembly threads for its [web
```

File: src/site/content/en/blog/webassembly-threads/index.md [hunk 20]
```diff
@@ -0,0 +1,472 @@
...
+*   [Loading the JS+Wasm generated by wasm-bindgen in a
+    Worker.](https://github.com/GoogleChromeLabs/wasm-bindgen-rayon/blob/e13485d6d64a062b890f5bb3a842b1fe609eb3c1/demo/wasm-worker.js#L28-L31)
+*   [Using wasm-bindgen-rayon to initialize a thread
+    pool.](https://github.com/GoogleChromeLabs/wasm-bindgen-rayon/blob/e13485d6d64a062b890f5bb3a842b1fe609eb3c1/demo/wasm-worker.js#L32)
+*   Using Comlink to [expose Worker's
+    API](https://github.com/GoogleChromeLabs/wasm-bindgen-rayon/blob/e13485d6d64a062b890f5bb3a842b1fe609eb3c1/demo/wasm-worker.js#L44-L46)
+    to [the main
+    thread](https://github.com/GoogleChromeLabs/wasm-bindgen-rayon/blob/e13485d6d64a062b890f5bb3a842b1fe609eb3c1/demo/index.js#L25-L29).
+
+## Real-world use-cases
```
Comment (mihajlija) [inline src/site/content/en/blog/webassembly-threads/index.md:456]: This is the interesting part, I suggest either adding a brief summary of exciting real world use cases to the intro or just moving the entire section at the top of the post. Always try to explain "why" before "how to".

File: src/site/content/en/blog/webassembly-threads/index.md [hunk 21]
```diff
@@ -0,0 +1,472 @@
...
+
+Google Earth is another notable service that's using WebAssembly threads for its [Web
+version](https://earth.google.com/web/).
+
+[FFMPEG.WASM](https://ffmpegwasm.github.io/) is a WebAssembly version of a popular
+[FFmpeg](https://www.ffmpeg.org/) multimedia toolchain that uses WebAssembly threads to efficiently
+encode videos directly in the browser.
+
+There are many more exciting examples using WebAssembly threads out there. Be sure to check out
+their demos and bring your own multithreaded applications and libraries to the Web!
```
Comment (mihajlija) [inline src/site/content/en/blog/webassembly-threads/index.md:472]:
```suggestion
their demos and bring your own multithreaded applications and libraries to the web!
```

File: src/site/content/en/blog/webassembly-threads/index.md [hunk 22]
```diff
@@ -0,0 +1,472 @@
...
+`SharedArrayBuffer`), precise timers, memory measurement and other APIs that require an isolated
+origin for security reasons. Check out the [linked guide](https://web.dev/coop-coep/) for more
+details.
+
+### WebAssembly atomics
+
+While `SharedArrayBuffer` allows each thread to read and write to the same memory, for correct
+communication you want to make sure they don't "step on each others' toes". For example, it's
+possible for one thread to start reading data from a shared address, while another thread is writing
+to it, so the first thread will now get a corrupted result. This category of bugs is known as race
```
[the remark this answers is not in the document]
Comment (RReverser) [reply src/site/content/en/blog/webassembly-threads/index.md:115]: This rewording changes a meaning, we definitely want threads acting on the same thing, but at the same time we want to make sure that whatever one thread is doing doesn't conflict with what another is doing while they are both touching the same thing.

File: src/site/content/en/blog/webassembly-threads/index.md [hunk 23]
```diff
@@ -0,0 +1,472 @@
...
+conditions. In order to prevent race conditions, you need to somehow synchronize those accesses.
+This is where atomic operations come in.
+
+[WebAssembly
+atomics](https://webassembly.github.io/threads/core/syntax/instructions.html#atomic-memory-instructions)
+is an extension to the WebAssembly instruction set that allow to read and write small cells of data
+(usually 32- and 64-bit integers) "atomically", that is, in a way that guarantees that no two
+threads are reading / writing to the same cell at the same time, preventing such conflicts at a low
+level. Additionally, WebAssembly atomics contain two more instruction kinds—"wait" and "notify"—that
```
[the remark this answers is not in the document]
Comment (RReverser) [reply src/site/content/en/blog/webassembly-threads/index.md:124]: I thought types might be a bit more ambigous, because it's not really new high-level type of instructions (like we have memory instructions, control flow instructions etc.). I guess it's ambigous either way and doesn't matter too much, but "kind" seems a bit more lightweight.

File: src/site/content/en/blog/webassembly-threads/index.md [hunk 24]
```diff
@@ -0,0 +1,472 @@
...
+### C
+
+In C, particularly on Unix-like systems, the common way to use threads is via [POSIX
+Threads](https://en.wikipedia.org/wiki/POSIX_Threads) provided by the `pthread` library. Emscripten
+[provides an API-compatible implementation](https://emscripten.org/docs/porting/pthreads.html) of
+the `pthread` library built atop Web Workers, shared memory and atomics, so that the same code can
+work on the Web without changes.
+
+Let's take a look at an example:
+
```
[the remark this answers is not in the document]
Comment (RReverser) [reply src/site/content/en/blog/webassembly-threads/index.md:170]: TIL, nice.

File: src/site/content/en/blog/webassembly-threads/index.md [hunk 25]
```diff
@@ -0,0 +1,472 @@
...
+```
+
+{% Aside 'caution' %} There is another problem though: see that `sleep(1)` in the code example? It
+executes in the thread callback, meaning off the main thread, so it should be fine, right? Well, it
+isn't. {% endAside %}
+
+When `pthread_join` is called, it has to wait for the thread execution to finish, meaning that if
+the created thread is performing long-running tasks—in this case, sleeping 1 second—then the main
+thread will also have to block for the same amount of time till the results are back. When this JS
+is executed in the browser, it will block the UI thread for 1 second until the thread callback
```
[the remark this answers is not in the document]
Comment (RReverser) [reply src/site/content/en/blog/webassembly-threads/index.md:286]: Should the "1 second" in the previous sentence be changed as well for symmetry?

File: src/site/content/en/blog/webassembly-threads/index.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (RReverser) [reply src/site/content/en/blog/webassembly-threads/index.md:281]: I agree, I also mentioned in the email that styling of asides didn't feel quite right here but thought maybe it's worth using it to highlight the problematic example anyway. I wonder if it's easier to just remove the aside again like it was in the draft...

File: src/site/content/en/blog/webassembly-threads/index.md (hunk 2 above)
Comment (RReverser) [reply src/site/content/en/blog/webassembly-threads/index.md:2]: Should it really duplicate the description? Is one derived from the other?

File: src/site/content/en/blog/webassembly-threads/index.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (RReverser) [reply src/site/content/en/blog/webassembly-threads/index.md:212]: Good point, thanks.

File: src/site/content/en/blog/webassembly-threads/index.md (hunk 20 above)
Comment (RReverser) [reply src/site/content/en/blog/webassembly-threads/index.md:456]: I explained "why" in the very beginning of the post, but specific examples feels like a "nice to have" part rather than something to lead with...

File: src/site/content/en/blog/webassembly-threads/index.md (hunk 18 above)
[the remark this answers is not in the document]
Comment (RReverser) [reply src/site/content/en/blog/webassembly-threads/index.md:398]: Hm, in this case technically it's just me. Should I use "I" instead or attribute it to Chrome or Chrome DevRel or ..?

File: src/site/content/en/blog/webassembly-threads/index.md (hunk 2 above)
Comment (mihajlija) [reply src/site/content/en/blog/webassembly-threads/index.md:2]: Description is for search engines, should be 1-3 sentence summary of the post. The subhead is what's displayed on the blog page, so it's a chance to give a bit more info than what fits in the title. They can be identical, though description could be more descriptive too. :)

File: src/site/content/en/blog/webassembly-threads/index.md (hunk 15 above)
Comment (RReverser) [reply src/site/content/en/blog/webassembly-threads/index.md:289]: I like this.

File: src/site/content/en/blog/webassembly-threads/index.md (hunk 16 above)
Comment (RReverser) [reply src/site/content/en/blog/webassembly-threads/index.md:395]: I somehow want to emphasise that the difference is really just a `par_` prefix that makes all of this to work, even though we're not manually creating and dealing with threads anywhere like we did in C/C++ examples.

Maybe rewording to "With this small change, the code ..." would be better?

File: src/site/content/en/blog/webassembly-threads/index.md (hunk 15 above)
Comment (RReverser) [reply src/site/content/en/blog/webassembly-threads/index.md:289]: Hm, need to think about how to link back to corresponding paragraphs though. Should those be subheadings?

File: src/site/content/en/blog/webassembly-threads/index.md (hunk 16 above)
Comment (mihajlija) [reply src/site/content/en/blog/webassembly-threads/index.md:395]: If the emphasis is important, then yes, that's better phrasing. Go for it 👍

File: src/site/content/en/blog/webassembly-threads/index.md (hunk 25 above)
[the remark this answers is not in the document]
Comment (mihajlija) [reply src/site/content/en/blog/webassembly-threads/index.md:286]: Actually, because this is a measurement it should be "1 second" per Google styleguide. @rachelandrew Here's the reference: https://developers.google.com/style/numbers?hl=en#numbers-as-numerals

File: src/site/content/en/blog/webassembly-threads/index.md (hunk 15 above)
Comment (mihajlija) [reply src/site/content/en/blog/webassembly-threads/index.md:289]: Yeah, just make sure those subheading have the correct level.

File: src/site/content/en/blog/webassembly-threads/index.md (hunk 20 above)
Comment (mihajlija) [reply src/site/content/en/blog/webassembly-threads/index.md:456]: Chatted with Rachel about this, you have a point, disregard the suggestion.

File: src/site/content/en/blog/webassembly-threads/index.md (hunk 22 above)
[the remark this answers is not in the document]
Comment (RReverser) [reply src/site/content/en/blog/webassembly-threads/index.md:116]: I'm going to close this one.

[the remark this answers is not in the document]
Comment (rachelandrew) [reply src/site/content/en/blog/webassembly-threads/index.md:116]: What I'm trying to avoid is the phrase "step on each other's toes" do you have a suggestion for something clearer?

File: src/site/content/en/blog/webassembly-threads/index.md (hunk 18 above)
[the remark this answers is not in the document]
Comment (RReverser) [reply src/site/content/en/blog/webassembly-threads/index.md:403]: Ping - any ideas?

File: src/site/content/en/blog/webassembly-threads/index.md [hunk 26]
```diff
@@ -0,0 +1,472 @@
...
+main thread alive for asynchronous event operations.
+Pthread 0xa05c18 exited.
+```
+
+### Rust
+
+Unlike Emscripten, Rust doesn't have a specialised end-to-end Web target, but instead provides a
+generic `wasm32-unknown-unknown` target for generic WebAssembly output.
+
+If Wasm is intended to be used in a Web environment, any interaction with JavaScript APIs is left to
```
Comment (mihajlija) [inline src/site/content/en/blog/webassembly-threads/index.md]:
```suggestion
If Wasm is intended to be used in a web environment, any interaction with JavaScript APIs is left to
```

File: src/site/content/en/blog/webassembly-threads/index.md (hunk 18 above)
[the remark this answers is not in the document]
Comment (mihajlija) [reply src/site/content/en/blog/webassembly-threads/index.md:403]: If you think the attribution is important in this case, you can do write something like "To [make the developer experience better] I've built the wasm-bindgen-rayon adapter which [does something]. 

If not, you can rewrite the sentence to omit attribution. Something like "You can access those hooks from wasm-bindgen-rayon adapter..."

File: src/site/content/en/blog/webassembly-threads/index.md (hunk 26 above)
Comment (RReverser) [reply src/site/content/en/blog/webassembly-threads/index.md]: Looks like this comment was left on outdated text, it already has "web" lowercase in latest version.

File: src/site/content/en/blog/webassembly-threads/index.md (hunk 15 above)
Comment (RReverser) [reply src/site/content/en/blog/webassembly-threads/index.md]: Done

File: src/site/content/en/blog/webassembly-threads/index.md (hunk 13 above)
Comment (RReverser) [reply src/site/content/en/blog/webassembly-threads/index.md:216]: Both C and C++ sections are all about Emscripten, so I don't think another heading is necessary here.

File: src/site/content/en/blog/webassembly-threads/index.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (RReverser) [reply src/site/content/en/blog/webassembly-threads/index.md:284]: I went forth and back on those, and while some highlighting seemed useful, I just couldn't come up with any usage of Asides here that actually helps rather than distracting from the flow, so on the 2nd thought just removed them again.

File: src/site/content/en/blog/webassembly-threads/index.md (hunk 12 above)
Comment (RReverser) [reply src/site/content/en/blog/webassembly-threads/index.md:220]: As mentioned below, I've just removed asides for now as I can't come up with a good non-distracting usage in this post.

File: src/site/content/en/blog/webassembly-threads/index.md (hunk 22 above)
[the remark this answers is not in the document]
Comment (rachelandrew) [reply src/site/content/en/blog/webassembly-threads/index.md:116]: @RReverser I'd like to find an alternative to "step on each other's toes" it is a colloquialism and something of an anthropomorphism (https://developers.google.com/style/anthropomorphism) given that threads don't have toes :) . This is what I was trying to get at, do you have a more descriptive way of saying this?

[the remark this answers is not in the document]
Comment (RReverser) [reply src/site/content/en/blog/webassembly-threads/index.md:116]: Ah. I Love adding some expressions like this, it feels more... human I guess. But you're right, it might not be widely understood and goes against the style. I can do this:

```suggestion
While `SharedArrayBuffer` allows each thread to read and write to the same memory, for correct
communication you want to make sure they don't perform conflicting operations at the same time. For example, it's
possible for one thread to start reading data from a shared address, while another thread is writing
to it, so the first thread will now get a corrupted result. This category of bugs is known as race
```

File: src/site/content/en/blog/webassembly-threads/index.md [hunk 27]
```diff
@@ -0,0 +1,497 @@
+---
+title: Using WebAssembly threads from C, C++ and Rust
+subhead: Learn how to bring multithreaded applications written in other languages to WebAssembly.
+description: Learn how to bring multithreaded applications written in other languages to WebAssembly.
+date: 2021-06-22
```
Comment (rachelandrew) [inline src/site/content/en/blog/webassembly-threads/index.md:5]:
```suggestion
date: 2021-07-08
```

**Review by rachelandrew — APPROVED** [review]
Approving so you should be able to merge (let me know if you can't and I will)

**Comment by RReverser** [comment]
Awesome, thanks! I'll merge tomorrow.

**Comment by RReverser** [comment]
(actually no, holding for next week to avoid publishing on Friday)
